### PR TITLE
KAFKA-3159: stale high watermark segment offset causes early fetch return

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

**Comment by hachikuji** [comment]
cc @junrao @becketqin @ijuma 

This is a follow-up to #688 which fixes the case when there are is only one replica in ISR.

**Comment by ijuma** [comment]
Like @junrao said in #688, I wonder if we should only compare message offset in `precedes` (as it used to be). And maybe we should rename it to `messageOffsetPrecedes` to make it unambiguous.

If we do that, then I think we also have to change the following in `Partition.maybeIncrementLeaderHW` to also check the `offsetOnOlderSegment` (to restore Becket's fix):

`if(oldHighWatermark.precedes(newHighWatermark)) {`

It would also be good to add a comment above Jason's change with some of Jun's explanation:

"The problem that I am seeing is that when there is only 1 replica in ISR, the only time that we will update HW is when a new message is appended. So, if no new message is appended and we roll a log segment, HW will stay on the old segment. Now, if a new fetch request comes in with log end offset, log.fetch() will return the new segment in LogOffsetMetadata. Then, in DelayedFetch.tryComplete(), since the fetch offset is not on the same segment as HW (even though offset is the same), we will complete the fetch request immediately and therefore incorrectly ignore max wait time."

Finally, a new JIRA for changing it so that HW moves to the new segment.

Thoughts?

**Comment by hachikuji** [comment]
@ijuma There's actually only two usages of the precedes() function currently. The first is used in `Partition` to determine if the high watermark should be updated. Checking the segment base offset makes sense to me in this case since we should update the high watermark when the segment changes. 

The other case is in `DelayedFetch` and the segment base offset check doesn't make as much sense since there is already an explicit prior check to handle differing segment offsets.

So maybe we could revert the precedes() change and just modify the check in `maybeIncrementLeaderHW` to check the segment base offset explicitly (e.g. with `LogOffsetMetadata.offsetOnOlderSegment`). What do you think?

**Comment by hachikuji** [comment]
@ijuma Haha, I think I suggested the same thing you already proposed. I guess I'll go ahead and make this change.

**Comment by ijuma** [comment]
Sounds good. :)

File: core/src/main/scala/kafka/server/LogOffsetMetadata.scala [hunk 1]
```diff
@@ -51,25 +51,15 @@ case class LogOffsetMetadata(messageOffset: Long,
...
-        false
-      else if (!this.messageOffsetOnly() && !that.messageOffsetOnly())
-        this.segmentBaseOffset < that.segmentBaseOffset
-      else
-        throw new KafkaException(s"Cannot compare $this with $that as one has segment base offsets and the other does not.")
-    }
-  }
-
+  // check if the message offset is before the given offset.
+  def precedes(that: LogOffsetMetadata): Boolean = this.messageOffset < that.messageOffset
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/LogOffsetMetadata.scala]: Do you think we should name this `messageOffsetPrecedes` to make it unambiguous?

File: core/src/main/scala/kafka/server/DelayedFetch.scala [hunk 2]
```diff
@@ -81,18 +81,20 @@ class DelayedFetch(delayMs: Long,
...
-              return forceComplete()
-            } else if (fetchOffset.offsetOnOlderSegment(endOffset)) {
-              // Case C, this can happen when the fetch operation is falling behind the current segment
-              // or the partition has just rolled a new segment
-              debug("Satisfying fetch %s immediately since it is fetching older segments.".format(fetchMetadata))
-              return forceComplete()
-            } else if (fetchOffset.precedes(endOffset)) {
-              // we need take the partition fetch size as upper bound when accumulating the bytes
-              accumulatedSize += math.min(endOffset.positionDiff(fetchOffset), fetchStatus.fetchInfo.fetchSize)
+            if (endOffset.messageOffset != fetchOffset.messageOffset) {
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/DelayedFetch.scala:87]: Probably worth adding a comment explaining this, as it's not obvious why we need it.

Comment (hachikuji) [inline core/src/main/scala/kafka/server/DelayedFetch.scala:87]: Agreed. Forgot to add this.

File: core/src/main/scala/kafka/server/LogOffsetMetadata.scala (hunk 1 above)
Comment (hachikuji) [inline core/src/main/scala/kafka/server/LogOffsetMetadata.scala]: Yeah, either that or replace it with the explicit check. Either way reads a little awkward:

``` scala
thisOffset.messageOffset < thatOffset.messageOffset
```

vs.

``` scala
thisOffset.messageOffsetPrecedes(thatOffset)
```

Either way probably works for me.

Comment (ijuma) [inline core/src/main/scala/kafka/server/LogOffsetMetadata.scala]: I agree and I'd probably go with option 1, but either is fine with me too. The fact that we've had at least 2 separate commits where the meaning of `precedes` wasn't well understood tells me that it's not a name we want to keep, even though it reads much more nicely. :)

**Comment by ijuma** [comment]
LGTM

File: core/src/main/scala/kafka/server/DelayedFetch.scala [hunk 3]
```diff
@@ -81,18 +81,23 @@ class DelayedFetch(delayMs: Long,
...
-            } else if (fetchOffset.offsetOnOlderSegment(endOffset)) {
-              // Case C, this can happen when the fetch operation is falling behind the current segment
-              // or the partition has just rolled a new segment
-              debug("Satisfying fetch %s immediately since it is fetching older segments.".format(fetchMetadata))
-              return forceComplete()
-            } else if (fetchOffset.precedes(endOffset)) {
-              // we need take the partition fetch size as upper bound when accumulating the bytes
-              accumulatedSize += math.min(endOffset.positionDiff(fetchOffset), fetchStatus.fetchInfo.fetchSize)
+            // Go directly to Case D checking if the message offsets are the same. This gets around the case where
+            // the current log segment has just rolled and the high watermark is still on the old segment,
```
Comment (junrao) [inline core/src/main/scala/kafka/server/DelayedFetch.scala]: Could we make it clear that this is the case when the segment has rolled, but the offset in HW remains the same?

**Comment by hachikuji** [comment]
@junrao Thanks for reviewing. I fixed the comment.

**Comment by junrao** [comment]
Thanks for the patch. LGTM

**Pull request closed without merging**